File: julia/4_contraction.jl
#!/usr/bin/env julia
"""
Tensor contraction examples (Julia)
"""

using Random
using TensorOperations

function main1()
    println("matrix-matrix multiplication")
    A = rand(2, 3)
    B = rand(3, 4)
    println("A: shape ", size(A), "\n", A, "\n")
    println("B: shape ", size(B), "\n", B, "\n")
    println("contract: A, B -> C")
    @tensor C[i, k] := A[i, j] * B[j, k]
    println("C: shape ", size(C), "\n", C, "\n")
end

function main2()
    println("more complex contraction")
    A = rand(2, 3, 4, 5)   # i j l m
    B = rand(4, 3)         # l n
    C = rand(5, 3, 4)      # m n k
    println("A: shape ", size(A))
    println("B: shape ", size(B))
    println("C: shape ", size(C))
    println("contract: A, B, C -> D")
    @tensor D[i, j, k] := A[i, j, l, m] * B[l, n] * C[m, n, k]
    println("D: shape ", size(D), "\n", D, "\n")
end


main1()
main2()

File: rust/src/bin/4_contraction.rs
use ndarray::Array;
use ndarray_einsum::einsum;
use rand::random;
use tn_basics::MapStrToAnyhowErr;

// Tensor contraction examples

fn main() -> anyhow::Result<()> {
    println!("matrix-matrix multiplication");
    let a = Array::from_shape_fn((2, 3), |_| random::<f64>());
    let b = Array::from_shape_fn((3, 4), |_| random::<f64>());
    println!("A: shape {:?}\n{}\n", a.shape(), a);
    println!("B: shape {:?}\n{}\n", b.shape(), b);
    println!("contract: A, B -> C");
    let c = einsum("ij,jk->ik", &[&a, &b]).map_str_err()?;
    println!("C: shape {:?}\n{}\n", c.shape(), c);

    println!("more complex contraction");
    let a = Array::from_shape_fn((2, 3, 4, 5), |_| random::<f64>());
    let b = Array::from_shape_fn((4, 3), |_| random::<f64>());
    let c = Array::from_shape_fn((5, 3, 4), |_| random::<f64>());
    println!("A: shape {:?}\n{}\n", a.shape(), a);
    println!("B: shape {:?}\n{}\n", b.shape(), b);
    println!("C: shape {:?}\n{}\n", c.shape(), c);
    println!("contract: A, B, C -> D");
    let d = einsum("ijlm,ln,mnk->ijk", &[&a, &b, &c]).map_str_err()?;
    println!("D: shape {:?}\n{}\n", d.shape(), d);
    Ok(())
}
